### PR TITLE
Support custom registry configuration

### PR DESCRIPTION
**Pull request by lllwan**

### CODE REVIEW

File: runtime/registry.go [hunk 1]
```diff
@@ -16,21 +16,56 @@ package runtime
...
+}
+
+func (d *Default) getRegistryConfig() *RegistryConfig {
+	var config RegistryConfig
+	registryConfigPath := filepath.Join(common.DefaultClusterBaseDir(d.ClusterName), "etc/registry.yaml")
+	if utils.IsFileExist(registryConfigPath) {
+		err := utils.UnmarshalYamlFile(registryConfigPath, &config)
+		if err == nil && config.IP != "" && config.Domain != "" {
+			if config.Volume == "" {
+				config.Volume = "/var/lib/registry"
```
Comment (fanux) [inline runtime/registry.go:49]: Volume=filepath.Join(DefaultClusterBaseDir(name),"/registry")

File: apply/cloud_applier.go [hunk 2]
```diff
@@ -43,6 +45,7 @@ func NewAliCloudProvider(cluster *v1.Cluster) Interface {
 		ImageManager:   image.NewImageService(),
 		FileSystem:     filesystem.NewFilesystem(),
 		Guest:          guest.NewGuestManager(),
+		Config:         config.NewConfiguration(cluster.Name),
```
Comment (fanux) [inline apply/cloud_applier.go:48]: cloud mod using config is a litter complexity，need copy the Clusterfile configs to the remote server, I will do it later, delete this now.

File: runtime/registry.go [hunk 3]
```diff
@@ -16,21 +16,56 @@ package runtime
...
+	}
+	return fmt.Sprintf("%s %s", cf.IP, cf.Domain)
+}
+
+func (d *Default) getRegistryConfig() *RegistryConfig {
+	var config RegistryConfig
+	registryConfigPath := filepath.Join(common.DefaultClusterBaseDir(d.ClusterName), "etc/registry.yaml")
+	if utils.IsFileExist(registryConfigPath) {
+		err := utils.UnmarshalYamlFile(registryConfigPath, &config)
+		if err == nil && config.IP != "" && config.Domain != "" {
```
Comment (fanux) [inline runtime/registry.go:47]: I think we can set default domain and default host IP here.

File: apply/cloud_applier.go (hunk 2 above)
Comment (lllwan) [reply apply/cloud_applier.go]: OK

File: runtime/registry.go (hunk 3 above)
Comment (lllwan) [reply runtime/registry.go:46]: default host IP defined in the getregistryhost() function. so, I don't think we need to change it.

File: runtime/registry.go (hunk 1 above)
Comment (lllwan) [reply runtime/registry.go]: If you need to do this, then the custom volume is meaningless. First remove the function of the custom volume.

File: runtime/registry.go [hunk 4]
```diff
@@ -16,21 +16,58 @@ package runtime
...
+	if cf == nil {
+		return fmt.Sprintf("%s %s", d.Masters[0], SeaHub)
+	}
+	return fmt.Sprintf("%s %s", cf.IP, cf.Domain)
+}
+
+func (d *Default) getRegistryConfig() *RegistryConfig {
+	var config RegistryConfig
+	registryConfigPath := filepath.Join(common.DefaultClusterBaseDir(d.ClusterName), "etc/registry.yaml")
+	if utils.IsFileExist(registryConfigPath) {
```
Comment (fanux) [inline runtime/registry.go:44]: if !utils.IsFileExist(registryConfigPath) {
   return nil
}

"return fast less retract"

Comment (lllwan) [reply runtime/registry.go:44]: OK

**Review by fanux — APPROVED** [review]

**Comment by fanux** [comment]
LGTM needs test

File: filesystem/filesystem.go [hunk 5]
```diff
@@ -147,6 +149,11 @@ func (c *FileSystem) UnMountRootfs(cluster *v1.Cluster) error {
 
 func mountRootfs(ipList []string, target string, cluster *v1.Cluster) error {
 	SSH := ssh.NewSSHByCluster(cluster)
+	d := runtime.Default{
```
Comment (fanux) [inline filesystem/filesystem.go:152]: Using the runtime struct seems not very good, other package only care about the runtime interface.

I think we can define the function GetRegistryConfig(defaultIP string), so you don't need the runtime.Default struct here.

Comment (lllwan) [reply filesystem/filesystem.go:152]: OK

File: config/config.go [hunk 6]
```diff
@@ -50,6 +50,7 @@ func NewConfiguration(clusterName string) Interface {
 
 func (c *Dumper) Dump(clusterfile string) error {
 	if clusterfile == "" {
+		logger.Warn("clusterfile is empty!")
```
Comment (fanux) [inline config/config.go:53]: If using run command clusterfile is "", but it normal, So I think using debug log here is better.

Comment (lllwan) [reply config/config.go]: OK

**Review by fanux — APPROVED** [review]

**Pull request merged**